kythe/go/platform/tools/kzip/mergecmd/mergecmd.go

```diff
@@ -57,7 +58,14 @@ func (c *mergeCommand) Execute(ctx context.Context, fs *flag.FlagSet, _ ...inter
 	if c.output == "" {
 		return c.Fail("required --output path missing")
 	}
-	if err := mergeArchives(ctx, c.output, fs.Args()); err != nil {
+	archives := fs.Args()
+	if len(archives) == 0 {
```

Comment:
Check that both STDIN and non-flag args are not given.

Reply:
i.e. you'd like a check like in all caps below

   if len(archives == 0) { // read from stdin }
   else { // read from args
      // *VERIFY STDIN IS EMPTY*
   }

?

Reply:
👍

Reply:
That makes it pretty annoying to run by hand from the shell (having to redirect stdin from devnull, or hitting ^D).

Reply:
You could make reading from stdin require a flag and then check that both that flag and `flag.NArg() != 0` are not true.  I'd just like to make it harder to accidentally lose inputs without any user notice.

The flag could be more generic as well.  It could point to a file to read lines from.  If you want stdin, pass `/dev/stdin`.

Reply:
Checked for terminal first.

Reply:
Ok, added a "xargs_mode" flag.

Reply:
I think I'd rather have a `--read_from_file` flag.  Special-casing `xargs` behavior seems weird.

Reply:
I think it's weird to have to specify /dev/stdin. If you want input from a file, use xargs_mode and redirect. That seems way more natural.

Reply:
If you just want the flag named "read_inputs_from_stdin" that's fine (just wordy)

Reply:
Is there some reason we can't _use_ `xargs` instead of adding a flag to this tool?

Reply:
To be more specific: Can we remove the restriction of one file per invocation, without getting ourselves into stdin-handling?

https://godoc.org/bitbucket.org/creachadair/fileinput might also help here, maybe.

Reply:
It's not possible to make xargs work in general. For any specific application, you need to specify an appropriate value for `-n` (max args per invocation of cmd) and `-s` (max size of cmdline per invocation). There's no (reasonable) values that will work in all circumstances.

If `kzip` actually read in the existing output file and merged into that, then xargs will work. I didn't want to go through that approach because, well, it was more work to write to a tmp output file, etc etc. It would also be poorly performing (n^2) as you re-merge the same data  for each invocation of kzip.

Reply:
I'm confused: It is perfectly possible to use `xargs` without specifying those things, provided the command-line works with its free arguments correctly. And if it doesn't—e.g., if we need a way to partition one batch of arguments from another—we shouldn't be using free arguments for that purpose IMO.

What am I missing?

Reply:
Perhaps more importantly: If there is a performance problem being addressed, what is the wider problem we are trying to solve here, and is there an issue/design doc around that talks about why we're doing this to solve it?

Reply:
The impetus for this came up because in [extract.sh](https://github.com/kythe/kythe/blob/7a1d39c7320fd5132e91fecb8679de1e36c823d6/kythe/extractors/bazel/extract.sh#L61) we essentially do `find ... | xargs kzip`. That fails because xargs picks defaults for `-n` and `-s`, and that ends up essentially doing
```
kzip --output o.kzip f1 f2 f3 ...
kzip --output o.kzip f100 f101 f202 ...
kzip --output o.kizp f200 f201 f202 ...
```
... which overwrites the output each invocation. Like I said, for any particular set of files I can choose `-n` and `-s` such that there is exactly one invocation of kzip, but I can't write a general purpose script to do it.

Reply:
What about allowing the `-output` flag to take a template with an (optional) randomizing string, like the way `mktemp` works?

Reply:
Or, another alternative: An `-output_dir` that, if set instead of `-output`, generates nonce output file names in that directory?

Both of these seem like they might be simpler to implement and less brittle to use.

Reply:
The other option, which I'm willing to consider, is to have kzip actually merge in the output file as well, if it exists. If we want to say "let's not worry about the performance of re-reading the same CUs and files multiple times", maybe OK, but I'd like to handle large-cardinality cases well.

Reply:
We want just one output file at the end of this; I'm not seeing how having an output_dir or a an output "template" helps here.  We'll then have to do 'multiphase' merging. Why can't the tool just support it?

Reply:
I don't think "simpler to implement" applies here. This is pretty simple. "Brittle", sure. I'll put in validation that either files to process come from stdin, or they appear as arguments.

Reply:
It appears you are trying to shoehorn a very specific use-case into a general-purpose tool, just to suit an existing script. I think a better solution would be to write a program that does what the script does, instead of polluting the API of a tool that doesn't suit that one use case.

Reply:
I'm not sure I agree. The use case is not all that specific: I have a lot (say 100k) of kzips, and I want to merge them. How do I do that? I'm proposing "let `kzip merge`" do that.  If `kzip merge` can't do that, then it's not very good merge tool; it's just a toy. I don't the high cardinality case is all that specific.

What do you think about the other proposal: let `kzip merge` preserve the existing output file contents, and essentially "append" to it (possibly controlled by some flag)?

Reply:
It's specific in that there is no good reason to combine hundreds of thousands of kzip files into one. The fact that we are doing so in this case appears to be an artifact of a particular way this script was set up. So my argument is that as a merge tool it already works fine: If you really need to be able to merge "into" an existing file, you could simply add that feature, though as you pointed out the ZIP format isn't well-situated for that case and you might want to consider a different format tif there's truly a need for huge single-file combinations.

I am skeptical that this is a real problem in need of solving, though.